cpp/src/io/utilities/type_inference.cuh

```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <io/utilities/column_type_histogram.hpp>
+#include <io/utilities/parsing_utils.cuh>
+#include <io/utilities/trie.cuh>
+
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_scalar.hpp>
+
+#include <thrust/distance.h>
+#include <thrust/tuple.h>
+
+#include <cstddef>
+
+namespace cudf {
+namespace io {
+namespace detail {
+/**
+ * @brief Non-owning view for type inference options
+ */
+struct inference_options_view {
+  cudf::detail::trie_view trie_true;
+  cudf::detail::trie_view trie_false;
+  cudf::detail::trie_view trie_na;
+  char quote_char;
+};
+/**
+ * @brief Structure for type inference options
+ */
+struct inference_options {
```

Comment:
Good point. I've removed this struct and create a `json_view()` member in `parse_options` returning `json_parse_options_view`.

Reply:
Thanks, @PointKernel. I think with that change in, we'll also want to:
1. Have a separate `json_parse_options` 
2. Make `json_view()` a member function of the new `json_parse_options` 
3. Migrate nested JSON's type casting to use `json_parse_options_view`

I will take care of these in a follow-up PR. Just want to get your thoughts on the procedure. Would allow us to have a clearer idea of options relevant/supported in JSON versus CSV.
cc: @vuule

Reply:
If the goal is to have a generic type inference functionality, why do we need a json-specific parse options?

Reply:
If the intended use for your structure is indeed just inference, we should reflect that in the class' name and call it something like `json_inference_options_view`. 

Otherwise, I was sort of suggesting to have a completely different struct `json_parse_options` that will _only_ reflect the options relevant for JSON. I'd like to hear @vuule 's thoughts on this first though.

Reply:
Renamed `json_parse_options_view` as `json_inference_options_view`